sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/ProjectRoleRepository.java

```diff
@@ -1,41 +1,170 @@
+
+import org.jetbrains.annotations.NotNull;
 
 import java.io.IOException;
 import java.util.EnumSet;
 import java.util.Set;
 import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
 
 /**
  * Repository for ProjectRoleBinding-based privileges.
  */
-public interface ProjectRoleRepository {
+public abstract class ProjectRoleRepository {
 
   /**
    * Find projects that a user has standing, requester privileges in.
    */
-  SortedSet<ProjectId> findProjectsWithRequesterPrivileges(
-      UserId user) throws AccessException, IOException;
+  abstract SortedSet<ProjectId> findProjectsWithRequesterPrivileges(
+      UserEmail user) throws AccessException, IOException;
 
   /**
    * List requester privileges for the given user.
    */
-  RequesterPrivilegeSet<ProjectRoleBinding> findRequesterPrivileges(
-      UserId user,
+  abstract RequesterPrivilegeSet<ProjectRoleBinding> findRequesterPrivileges(
+      UserEmail user,
       ProjectId projectId,
       Set<ActivationType> typesToInclude,
       EnumSet<RequesterPrivilege.Status> statusesToInclude) throws AccessException, IOException;
 
   /**
    * List users that hold an eligible reviewer privilege for a role binding.
    */
-  Set<UserId> findReviewerPrivelegeHolders(
+  abstract Set<UserEmail> findReviewerPrivelegeHolders(
       ProjectRoleBinding roleBinding,
       ActivationType activationType) throws AccessException, IOException;
+
+  static <T extends PrivilegeId> @NotNull RequesterPrivilegeSet<T> buildRequesterPrivilegeSet(
+      @NotNull Set<RequesterPrivilege<T>> availableRequesterPrivileges,
+      @NotNull Set<ActivatedRequesterPrivilege<T>> validActivations, // TODO(later): rename to active
+      @NotNull Set<ActivatedRequesterPrivilege<T>> expiredActivations,
+      @NotNull Set<String> warnings) {
+    assert availableRequesterPrivileges.stream().allMatch(e -> e.status() == RequesterPrivilege.Status.INACTIVE);
+    assert validActivations.stream().noneMatch(id -> expiredActivations.contains(id));
+    assert expiredActivations.stream().noneMatch(id -> validActivations.contains(id));
+
+    //
+    // Return a set containing:
+    //
+    // 1. Available privilegges
+    // 2. Active privileges
+    //
+    // where (1) and (2) don't overlap.
+    //
+    // Expired privileges are ignored.
+    //
+    var availableAndInactive = availableRequesterPrivileges
+        .stream()
+        .filter(privilege -> validActivations
+            .stream()
+            .noneMatch(active -> active.privilegeId().equals(privilege.id())))
+        .collect(Collectors.toCollection(TreeSet::new));
+
+    assert availableAndInactive.stream().noneMatch(e -> validActivations.contains(e.id()));
+
+    var current = new TreeSet<RequesterPrivilege<T>>(availableAndInactive);
+    for (var validActivation : validActivations) {
+      //
+      // Find the corresponding privilege to determine
+      // whether this is JIT or MPA-eligible.
+      //
```

Comment:
Does this part also take into account your changes?

Reply:
Yes this is a mix of my changes and changes to main repo.